feat(hd_wallet): evm hd wallet and trezor

shamardy opened the pull request:
Part of https://github.com/KomodoPlatform/komodo-defi-framework/issues/1838

This is an in progress PR:
- [x] Refactor HD wallet functionalities to be separate from utxo so it can be used easily for other coins.
- [x] Add `HDWallet` derivation to `EthCoin`.
- [x] Implement `HDWalletCoinOps`, `HDWalletBalanceOps`, etc.. for `EthCoin`.
- [x] Balance streaming for all addresses https://github.com/KomodoPlatform/komodo-defi-framework/pull/2041#discussion_r1462246644
- [x] Implement `init_platform_coin_with_tokens` using task manager for `EthCoin`.
- [x] Implement `init_token` using task manager.
- [x] Maintain backward compatibility.
- [x] Doc comments

Changes:
- EVM balance streaming now contains address field to support HD wallet
- `chain_id` is needed for all evm coins/tokens. `chain_id` should be part of platform coin config only actually and retrieved from platform coin, this is for v2 methods, but we should keep it in all coins config for v1 methods.

artemii235 reviewed (changes requested):
Great work!
It seems that few traits can be removed, which will hopefully make code simpler 🙂

shamardy commented on mm2src/coins/hd_wallet/mod.rs:
```diff
@@ -0,0 +1,431 @@
+    pub derivation_path: DerivationPath,
+}
+
+impl<Address, Pubkey> HDAddressOps for HDAddress<Address, Pubkey>
+where
+    Address: Clone + Display + Send + Sync,
+{
+    type Address = Address;
+    type Pubkey = Pubkey;
+
+    fn address(&self) -> &Self::Address { &self.address }
+
+    fn pubkey(&self) -> &Self::Pubkey { &self.pubkey }
+
+    fn derivation_path(&self) -> &DerivationPath { &self.derivation_path }
+}
```
Thank you for bringing this point up to discussion, I decided to add `HDAddressOps` for 2 reasons:

1 - While the `HDAddress` generic struct serves as a general-purpose representation suitable for almost all coins, some coins in the future might require additional fields. An example would be `ZCoin` where it might be good to implement this trait for a slightly different `ZcoinHDAddress` Struct that includes and additional field for `my_z_addr_encoded`. https://github.com/KomodoPlatform/komodo-defi-framework/blob/6b751b865b7a3020471fd06ce4e80c6d1ccffc65/mm2src/coins/z_coin.rs#L202
Zcoin will probably need a different struct from the generic `HDAccount` too depending how we handle this case https://github.com/KomodoPlatform/komodo-defi-framework/pull/1933#discussion_r1293013105, but the generic `HDWallet` will probably be reused for `Zcoin` where we plug `ZcoinHDAccount` into it.

2 - Another reason was to design a crate for the framework/SDK, SDKs crates best practices often involve providing traits with a generic struct that implements these traits, this allows the users of such crates to opt out of using the generic struct if they want, an example would be a user that wants to only keep the pubkey inside their custom `HDAddress` struct and use `fn address(&self) -> &Self::Address` to derive the address from the pubkey.

I am not completely against removing `HDAddressOps` but wanted to provide some context into the thinking process that went into it, if you still think it's good to remove it to reduce boilerplate code I won't object :)

artemii235 replied:
These are great points, so I don't think that it's good to remove this and other related traits anymore, thanks 🙂 

I have one note regarding
> use fn address(&self) -> &Self::Address to derive the address from the pubkey.

Since the method returns reference, I think it won't be possible to return the derived address because the value will exist only in the method body.

shamardy replied:
> Since the method returns reference, I think it won't be possible to return the derived address because the value will exist only in the method body.

Fixed the method to return a non-reference.

shamardy commented on mm2src/coins/hd_wallet/errors.rs:
```diff
@@ -0,0 +1,156 @@
+}
+
+#[derive(Display)]
+pub enum NewAccountCreatingError {
```
Done

artemii235 commented on mm2src/coins/coin_balance.rs:
```diff
@@ -15,19 +13,20 @@ use std::ops::Range;
 pub enum EnableCoinBalanceError {
     NewAddressDerivingError(NewAddressDerivingError),
-    NewAccountCreatingError(NewAccountCreatingError),
+    NewAccountCreatingError(NewAccountCreationError),
```
It's worth changing `Creating` to `Creation` here too 🙂

shamardy replied:
done

artemii235 commented on mm2src/coins/hd_wallet/mod.rs:
```diff
@@ -0,0 +1,431 @@
+    }
+
+    async fn get_enabled_address(&self) -> Option<HDWalletAddress<Self>> {
+        let enabled_address = match self.enabled_address.clone() {
```
`?` can be used here:
```rust
let enabled_address = self.enabled_address.clone()?;
```

shamardy replied:
done

artemii235 commented on mm2src/coins/hd_wallet/mod.rs:
```diff
@@ -0,0 +1,431 @@
+            None => return None,
+        };
+
+        let account = match self.get_account(enabled_address.account_id).await {
```
`?` can be used here: 
```rust
self.get_account(enabled_address.account_id).await?;
```

shamardy replied:
done

artemii235 commented on mm2src/coins/hd_wallet/mod.rs:
```diff
@@ -0,0 +1,431 @@
+    fn hd_wallet_storage(&self) -> &HDWalletCoinStorage { &self.hd_wallet_storage }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
```
`Copy` can also be derived for `HDAccountAddressId`.

laruh replied:
> `Copy` can also be derived for `HDAccountAddressId`.

I have a question, why there is a need in `Copy` for  this struct?

artemii235 replied:
> Generally speaking, if your type can implement Copy, it should.

https://doc.rust-lang.org/std/marker/trait.Copy.html#when-should-my-type-be-copy

🙂

laruh replied:
Interesting, ok :D

shamardy replied:
done

artemii235 commented on mm2src/coins/hd_wallet/mod.rs:
```diff
@@ -0,0 +1,431 @@
+        account_der_path.push(ChildNumber::new(self.account_id, true)?);
+        account_der_path.push(self.chain.to_child_number());
+        account_der_path.push(ChildNumber::new(self.address_id, false)?);
+        drop_mutability!(account_der_path);
```
Is this really required for value that is immediately returned?

shamardy replied:
It's not required, I like to add it in case a function is extended later. But for this specific case, the `to_derivation_path` function appears to be fixed and will never be subject to further extension.

shamardy replied:
done

artemii235 reviewed (changes requested):
Few minor notes 🙂

artemii235 reviewed (approved):
👍

shamardy commented:
Thanks you for the review @artemii235, I will continue working through this list https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979#issue-1919982869 in this sprint by adding new commits to this PR, so next review in a few weeks should be for commits after this comment only to make it easier to review :)

dimxy commented on mm2src/coins/hd_wallet/mod.rs:
```diff
@@ -0,0 +1,515 @@
+use async_trait::async_trait;
```
I wonder why mod.rs was added instead hs_wallet.rs?
I can see the general trend now is to avoid mod.rs files to have more descriptive file names

shamardy replied:
I plan to make the `hd_wallet` directory a crate where I rename `mod.rs` to `lib.rs`. To make it a crate I need to remove any coin specific logic or structs from it, only `UtxoAddressFormat` is what's left then I can make it a separate crate.

shamardy commented:
For next review iteration that should start next week, please note that this is still a WIP PR. I left a lot of todos throughout the code to indicate some of what is left in addition to the list here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979#issue-1919982869.

shamardy commented:
@laruh @dimxy @onur-ozkan Please only review commits added since the last approval. I will still add a few more commits to this PR but they are mostly not related to functionality and more related to doc comments and tests.

@onur-ozkan you can review balance streaming fixes only and maybe take a look at trezor implementation for inspirations for external signers (if needed).

onur-ozkan commented:
> Please only review commits added since the last approval. I will still add a few more commits to this PR but they are mostly not related to functionality and more related to doc comments and tests.

Unfortunately that it is not so possible for this big PRs. If you try to make reviews commit-by-commit you will spend too much time on reviewing unrelated changes (that are removed or changed later) for no reason and it will make reviewing much harder/complicated.

dimxy commented:
> > Please only review commits added since the last approval. I will still add a few more commits to this PR but they are mostly not related to functionality and more related to doc comments and tests.
> 
> Unfortunately that it is not so possible for this big PRs. If you try to make reviews commit-by-commit you will spend too much time on reviewing unrelated changes (that are removed or changed later) for no reason and it will make reviewing much harder/complicated.

Maybe in future we could use PR-to-PR approach for big feature additions, to avoid dismissal of past reviews. Any problems regarding this approach?

onur-ozkan commented:
> Maybe in future we could use PR-to-PR approach for big feature additions, to avoid dismissal of past reviews. Any problems regarding this approach?

For this big PRs that will be also hard to track/review. Why not just implementing and merging things gradually? We don't need to merge everything at once.

shamardy commented:
> Unfortunately that it is not so possible for this big PRs. If you try to make reviews commit-by-commit you will spend too much time on reviewing unrelated changes (that are removed or changed later) for no reason and it will make reviewing much harder/complicated.

These are the changes since the last approval https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/files/4d6e468b6da7555215557045d2fbeda2885b79d7..ffd0c965d8d071e5f9d0e9f08cb5aa1847d3ac71 I am not sure if that helps or not. If this is not helpful, we can arrange on DM which parts to be covered by who to make the review easier.

> Maybe in future we could use PR-to-PR approach for big feature additions, to avoid dismissal of past reviews. Any problems regarding this approach?

We will not use PR-to-PR approach anymore, this PR and the related ones are leftovers from when we used to wait for QA and sec review approvals before merging PRs, to continue development we opened feature branches from other feature branches which caused a lot of problems and this PR is an example of that.

> Why not just implementing and merging things gradually?

That's our current approach, sorry for the confusion. I know that for my other open PR https://github.com/KomodoPlatform/komodo-defi-framework/pull/2014 I don't follow this but the seed generation PR is a special case where I wanted to finish everything related to it there and have sec-review done on the PR for a such critical implementation. But I will merge it after the next 2 weeks on its current state and continue in another PR to make things easier from now on.

onur-ozkan commented on mm2src/coins/eth/eth_balance_events.rs:
```diff
@@ -16,49 +17,79 @@ use super::EthCoin;
-        .collect::<FuturesUnordered<_>>();
-
-    jobs.collect().await
+    let addresses = match coin.my_addresses().await {
+        Ok(addresses) => addresses,
+        Err(e) => {
+            return vec![Err((coin.ticker.clone(), e.into()))];
+        },
+    };
```
When any of the account fails during the derivation, balance streaming on ETH will stop for all the addresses.

shamardy replied:
`my_addresses` will never fail. Error handling is added because of the reliance on `derive_child` from `bip32` crate here https://github.com/KomodoPlatform/komodo-defi-framework/blob/ffd0c965d8d071e5f9d0e9f08cb5aa1847d3ac71/mm2src/coins/hd_wallet/coin_ops.rs#L63-L66 but we never pass to it any values that will cause errors.

onur-ozkan replied:
I think we should add an error log saying about it (like this is an unexpected error or something).

onur-ozkan commented on mm2src/coins/eth/eth_balance_events.rs:
```diff
@@ -16,49 +17,79 @@ use super::EthCoin;
+        // type and mapping the platform coin and the entire token list (which can grow at any time), we map
+        // the platform coin to Erc20TokenInfo so that we can use the token list right away without
+        // additional mapping.
+        tokens.insert(coin.ticker.clone(), Erc20TokenInfo {
+            token_address: address,
+            decimals: coin.decimals,
+        });
```
Don't we have those addresses in the `erc20_tokens_infos` already? If not, I think we should if I am not missing something.

shamardy replied:
I followed what you did there https://github.com/KomodoPlatform/komodo-defi-framework/blob/96fac7c8c97b5c2267bfd9bd250237817269acfe/mm2src/coins/eth/eth_balance_events.rs#L27-L36 but I guess I made a mistake. We only need the platform ticker once, I guess I can use all zeros address for the token address of the platform coin and add it before the loop, will fix this.

shamardy replied:
Fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/37e2b7102152bc6655465b2c6c91df1876c4519f

onur-ozkan commented on mm2src/coins/eth/eth_balance_events.rs:
```diff
@@ -16,49 +17,79 @@ use super::EthCoin;
     token_ticker: String,
     info: &Erc20TokenInfo,
-) -> Result<(String, BigDecimal), (String, MmError<BalanceError>)> {
+) -> Result<(String, (String, BigDecimal)), (String, MmError<BalanceError>)> {
```
I think it's time to create a type for this. :smile: At the first look I had no idea what the extra String is used for.

shamardy replied:
Will create a type for sure. I am conflicted about whether we should have the balance data be ticker-based or address-based. I guess having it ticker-based, as it is currently done, while adding the total balance across all addresses, will be the best solution. It will also align with how GUIs are more coin-based than address-based currently. However, this might change depending on how GUIs implement HD wallets.

shamardy replied:
> I guess having it ticker-based, as it is currently done, while adding the total balance across all addresses, will be the best solution

Nevermind this, we only broadcast if a balance changes for an address so no need for total balance.

shamardy replied:
fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/86603926a2972fb94195f8693dd144e10c5bed13

shamardy commented on mm2src/coins/coin_balance.rs:
```diff
@@ -33,59 +35,104 @@ impl From<BalanceError> for EnableCoinBalanceError {
+        Self: Sized;
+
+    /// Returns the total balance for the specified ticker.
+    fn get_total_for_ticker(&self, ticker: &str) -> Option<BigDecimal>;
```
Fixed

shamardy commented:
@onur-ozkan all comments should be fixed except this https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979#discussion_r1538934870 where I posted a reply.

onur-ozkan commented on mm2src/coins/eth/eth_balance_events.rs:
```diff
@@ -15,60 +16,106 @@ use std::collections::HashMap;
+
+    let addresses = match coin.my_addresses().await {
+        Ok(addresses) => addresses,
+        Err(e) => {
+            return vec![Err(BalanceFetchError {
+                ticker: coin.ticker.clone(),
+                address: "".to_string(),
+                error: e.into(),
+            })];
+        },
+    };
```
I think it's better to just unwrap it if it never fails? Without an address the error doesn't make sense since we propagate the error to the client using both the address and ticker.

shamardy replied:
> I think it's better to just unwrap it if it never fails?

It should never fail, but if we used unwrap, if a dev made an error that is missed in a review it will lead to panicking. I will add an additional error log as you suggested here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979#discussion_r1540476735
I can use `my_addresses` outside `get_all_balance_results_concurrently` instead and pass them to `get_all_balance_results_concurrently`, if `my_addresses` fails we can send the error as an event which will not have address in it.

shamardy replied:
> I can use my_addresses outside get_all_balance_results_concurrently instead and pass them to get_all_balance_results_concurrently, if my_addresses fails we can send the error as an event which will not have address in it.

Fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/846c22a70ab2f62d368908f0e383681679cd1736

onur-ozkan commented on mm2src/coins/eth/eth_balance_events.rs:
```diff
@@ -15,60 +16,106 @@ use std::collections::HashMap;
     // additional mapping.
     tokens.insert(coin.ticker.clone(), Erc20TokenInfo {
-        token_address: coin.my_address,
+        token_address: Address::default(),
```
With this approach HD account balance events will be unrecognizable from the clients. They need to know which address and the ticker balance was changed for.

shamardy replied:
This is the token contract address, if the coin is a platform coin, it doesn't have a token contract, if it's a token but `token_ticker == coin.ticker`, `address_balance` is used for balance
https://github.com/KomodoPlatform/komodo-defi-framework/blob/86603926a2972fb94195f8693dd144e10c5bed13/mm2src/coins/eth/eth_balance_events.rs#L83-L85

onur-ozkan replied:
So there are 2 issues:

1- fetch_balance may use `info.token_address` to fetch balance but it returns `address` at the end.
2- When you pass `Address::default()`, this value will be used for sending balance changes in the balance events.

shamardy replied:
I guess I don't understand the problem, maybe I am missing something. Here is my explanation of what is going on, please correct me if I am wrong

> 1- fetch_balance may use info.token_address to fetch balance but it returns address at the end.

`get_token_balance` takes the token contract address which is `info.token_address`, what is returned at the end is the user address, which is one of the addresses of the HD wallet passed to `get_all_balance_results_concurrently`.  `get_token_balance` is never used for this ticker https://github.com/KomodoPlatform/komodo-defi-framework/blob/846c22a70ab2f62d368908f0e383681679cd1736/mm2src/coins/eth/eth_balance_events.rs#L44-L47 previously (before this PR) you passed the user's address as the token address which was not right too but it was fine because it never gets used with `get_token_balance`.

> When you pass Address::default(), this value will be used for sending balance changes in the balance events. Here

`Address::default()` which is a dummy for platform coin contract address is never returned to the function sending the balance changes, the user address/s is what is returned.

shamardy replied:
> previously (before this PR) you passed the user's address as the token address which was not right too but it was fine because it never gets used with get_token_balance.

P.S. I can't pass the user's address as token contract address as before because we now have multiple addresses per user.

onur-ozkan replied:
I got confused with the `address` argument of `fetch_balance` and `address` field of `Erc20TokenInfo`.. I implemented this sometime ago now it looks very confusing at the first look. Perhaps adding explanation comment on top of `Erc20TokenInfo` address would be good for the team?

shamardy replied:
I agree that adding more comments will help, I too made a mistake before ref. https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979#discussion_r1539248928 because it wasn't very clear to me.

shamardy replied:
Added some comments that explain the situation here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/dc6ef2ae1f9d9acffc86e6c73943389bd27378c2

laruh commented on mm2src/coins_activation/src/eth_with_token_activation.rs:
```diff
@@ -157,39 +191,80 @@ impl RegisterTokenInfo<EthCoin> for EthCoin {
+pub struct HDEthWithTokensActivationResult {
+    current_block: u64,
+    ticker: String,
+    wallet_balance: CoinBalanceReport<CoinBalanceMap>,
+    // Todo: Move to wallet_balance when implementing HDWallet for NFTs
```
so nft info will be part of this structure?

```rust
/// Represents the balance of a single address in an HD wallet.
#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct HDAddressBalance<BalanceMap> {
    pub address: String,
    pub derivation_path: RpcDerivationPath,
    pub chain: Bip44Chain,
    pub balance: BalanceMap,
}
```

laruh reviewed (commented):
Impressive work!
I have questions and non blocking comments

dimxy commented on mm2src/coins/hd_wallet/confirm_address.rs:
```diff
@@ -83,7 +83,7 @@ where
     Task::UserAction: TryIntoUserAction + Send,
 {
-    async fn confirm_utxo_address(
+    async fn confirm_address(
```
I could not find that confirm_address() can work with trezor eth api although GetNewAddressRpcOps is implemented for EthCoin. is this a TODO?

shamardy replied:
Can you please explain this in more details, should I skip this https://github.com/KomodoPlatform/komodo-defi-framework/blob/ffd0c965d8d071e5f9d0e9f08cb5aa1847d3ac71/mm2src/coins/hd_wallet/coin_ops.rs#L240-L243 for EVM coins or implement `confirm_eth_address_with_trezor` function and add it to `confirm_address` implementation.
P.S. No this is not a TODO, Trezor and HD wallet for EVM should completely work after this PR is merged.

dimxy replied:
Yes, I can see that confim_address() implementation calls Self::confirm_utxo_address_with_trezor but I think this implementation also could be called for the eth coin

shamardy replied:
should be fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/d9777aff5afdfab54514a07a81c279c2695615fd

shamardy commented on mm2src/coins_activation/src/eth_with_token_activation.rs:
```diff
@@ -343,5 +466,21 @@ fn eth_priv_key_build_policy(
+
+pub type EthTaskManagerShared = InitPlatformCoinWithTokensTaskManagerShared<EthCoin>;
+
+pub(crate) fn eth_xpub_extractor_rpc_statuses(
+) -> HwConnectStatuses<InitPlatformCoinWithTokensInProgressStatus, InitPlatformCoinWithTokensAwaitingStatus> {
+    HwConnectStatuses {
+        on_connect: InitPlatformCoinWithTokensInProgressStatus::WaitingForTrezorToConnect,
+        on_connected: InitPlatformCoinWithTokensInProgressStatus::ActivatingCoin,
+        on_connection_failed: InitPlatformCoinWithTokensInProgressStatus::Finishing,
+        on_button_request: InitPlatformCoinWithTokensInProgressStatus::FollowHwDeviceInstructions,
+        on_pin_request: InitPlatformCoinWithTokensAwaitingStatus::EnterTrezorPin,
+        on_passphrase_request: InitPlatformCoinWithTokensAwaitingStatus::EnterTrezorPassphrase,
+        on_ready: InitPlatformCoinWithTokensInProgressStatus::ActivatingCoin,
```
It's similar to this for utxo
https://github.com/KomodoPlatform/komodo-defi-framework/blob/ef359cab467402370687979788887836939b68ae/mm2src/coins_activation/src/utxo_activation/common_impl.rs#L76-L87
where it's placed in the same file as it's used. I can move it to `platform_coin_with_tokens.rs` which is the common file for any implementation that implements `PlatformCoinWithTokensActivationOps`

shamardy replied:
fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/3c2e77d80b95bfb828ae53b6f1eca79fd7c01575

shamardy commented on mm2src/coins_activation/src/init_token.rs:
```diff
@@ -0,0 +1,353 @@
+    activation_params: T,
+}
+
+/// Trait for the initializing an EVM token using the task manager.
+#[async_trait]
+pub trait InitTokenActivationOps: Into<MmCoinEnum> + TokenOf + Clone + Send + Sync + 'static {
```
It's not only for EVM, will change the doc comment.

shamardy replied:
Fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/2a9e79073eea53a5958507e35d0d9c6b3083041d

shamardy commented on mm2src/coins_activation/src/init_erc20_token_activation.rs:
```diff
@@ -0,0 +1,197 @@
+
+pub(crate) fn erc20_xpub_extractor_rpc_statuses(
+) -> HwConnectStatuses<InitTokenInProgressStatus, InitTokenAwaitingStatus> {
+    HwConnectStatuses {
```
fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/7b812064457cfa7ceaaf59cb09598c1bdec4e664

dimxy commented on mm2src/trezor/src/eth/eth_command.rs:
```diff
@@ -0,0 +1,214 @@
+}
+
+impl<'a> TrezorSession<'a> {
+    /// Unused for now. Added for future use like implementation of an rpc returning eth address
```
Plz fix this note that this fn is unused. This is not true now

shamardy replied:
Fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/ee8ecf34248247a332045f4170d60766a2fe3e7f

dimxy commented on mm2src/coins/rpc_command/account_balance.rs:
```diff
@@ -27,51 +26,71 @@ pub struct AccountBalanceParams {
 
 #[derive(Debug, PartialEq, Serialize)]
-pub struct HDAccountBalanceResponse {
+pub struct HDAccountBalanceResponse<BalanceMap> {
```
`BalanceMap` is not necessary a map. Maybe just `Balance` or `BalanceHolder`?

shamardy replied:
How about `BalanceObject`?

shamardy replied:
Fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/09c12606641d9bcb6799ae6bb1d7b35e37f74169

laruh reviewed (commented):
I have a question

laruh commented on mm2src/coins/hd_wallet/account_ops.rs:
```diff
@@ -14,6 +14,23 @@ use mm2_err_handle::prelude::*;
 
+    /// Creates a new `HDAccountOps` instance.
+    ///
+    /// # Parameters
+    ///
+    /// * `account_id`: The index of the account.
+    /// * `account_extended_pubkey`: The extended public key associated with this account.
+    /// * `account_derivation_path`: The derivation path from the master key to this account.
+    ///
+    /// # Returns
+    ///
+    /// A new `HDAccountOps` instance.
```
Could you write doc comments above fields types (structures, pub types, enum etc), instead of function arguments repetition?
If some field needs specific note, you can this info as an additional sentence after `Creates a new `HDAccountOps` instance.`.
see https://github.com/KomodoPlatform/komodo-defi-framework/pull/2049#discussion_r1489331792 , https://github.com/KomodoPlatform/komodo-defi-framework/pull/2014#discussion_r1487411900

please do this for other functions

shamardy replied:
Thanks for the reminder, I guess this commit was before these comments https://github.com/KomodoPlatform/komodo-defi-framework/pull/2049#discussion_r1489331792 , https://github.com/KomodoPlatform/komodo-defi-framework/pull/2014#discussion_r1487411900

shamardy replied:
Fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/503aafbf99b03908a2f485ccfaf66880cb4a6474

dimxy commented on mm2src/coins/rpc_command/account_balance.rs:
```diff
@@ -26,48 +26,68 @@ pub struct AccountBalanceParams {
+        MmCoinEnum::QtumCoin(qtum) => Ok(HDAccountBalanceResponseEnum::Single(
+            qtum.account_balance_rpc(req.params).await?,
+        )),
+        MmCoinEnum::EthCoin(eth) => Ok(HDAccountBalanceResponseEnum::Map(
+            eth.account_balance_rpc(req.params).await?,
```
I also was going to ask: maybe use the same format for all coins, like for utxo just return a map of a single element, to simplify processing for the GUI ?

shamardy replied:
> Could you please clarify why we use HDAccountBalanceResponseEnum::Map for the eth coin, and Single for the other supported assets?

This is a map of coins/tokens ticker to balance, so we use map for ETH/EVM because the response should contain for each address the platform coin balance and the tokens balances. Other coins don't have tokens currently, we can add tokens support for them in the future but we would use a new API method, for backward compatibility of current APIs we shouldn't return a map as this is used currently in GUIs (e.g. for UTXO trezor).

P.S. GUIs would need to change their code for already implemented functionalities if we used Map for previously implemented coins. It's not only GUIs but any one running the defi framework and uses these methods.

shamardy replied:
> maybe use the same format for all coins

I want to do that in the future, but like I said they have to be new APIs.

laruh replied:
>  we use map for ETH/EVM because the response should contain for each address the platform coin balance and the tokens balances. Other coins don't have tokens currently

Oh tokens, exactly

Please add doc comments for `pub enum HDAccountBalanceResponseEnum` and for `pub struct HDAccountBalanceResponse<BalanceObject>` about this

shamardy commented on mm2src/coins/test_coin.rs:
```diff
@@ -439,7 +447,7 @@ impl CoinAssocTypes for TestCoin {
     type SigParseError = String;
 
-    fn my_addr(&self) -> &Self::Address { todo!() }
+    async fn my_addr(&self) -> Self::Address { todo!() }
```
It wasn't named by me, but I think Artem wanted to distinguish between the 2 since this is supposed to be used in swaps v2 (trading protocol upgrade) context.

dimxy commented on mm2src/coins/eth/eth_balance_events.rs:
```diff
@@ -80,31 +120,69 @@ impl EventBehaviour for EthCoin {
                             ctx.stream_channel_controller
                                 .broadcast(Event::new(
-                                    format!("{}:{}", EthCoin::ERROR_EVENT_NAME, ticker),
+                                    format!("{}:{}:{}", EthCoin::ERROR_EVENT_NAME, err.ticker, err.address),
```
Wondering why we did not create a enum for event names as a common registry? As I can see, we duplicate event names in different coins. Overlaps are also possible if no single registry

shamardy replied:
Since we started with an implementation for one coin and added other coins gradually I guess this wasn't considered, but it was considered in a recent review comment ref. https://github.com/KomodoPlatform/komodo-defi-framework/pull/2076#discussion_r1537603729
I can refactor it here but need to merge with latest dev first
c.c. @onur-ozkan

shamardy replied:
Fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/2d8b0e2cb9e4c8a7b438cce1f87d433e4caec2c3
c.c. @onur-ozkan @borngraced

onur-ozkan replied:
The format is good (I am using similar format for another event here [1](https://github.com/KomodoPlatform/komodo-defi-framework/blob/985997b47d84f9025e371be0f7e971e6bcdb0be5/mm2src/mm2_core/src/data_asker.rs#L72) [2](https://github.com/KomodoPlatform/komodo-defi-framework/blob/985997b47d84f9025e371be0f7e971e6bcdb0be5/mm2src/coins/tendermint/tendermint_coin.rs#L1051)), but we need a format consistency for per events (if we use this format, we should use it on all other `COIN_BALANCE` messages too.), otherwise it would be painful from clients to parse the event.

dimxy commented on mm2src/coins/lp_coins.rs:
```diff
@@ -2741,6 +2750,7 @@ impl HttpStatusCode for WithdrawError {
             | WithdrawError::ContractTypeDoesntSupportNftWithdrawing(_)
             | WithdrawError::CoinDoesntSupportNftWithdraw { .. }
             | WithdrawError::NotEnoughNftsAmount { .. }
+            | WithdrawError::ChainIdRequired(_)
```
I created this error but now I see it lacks a coin field instead of String field
Also possibly good to add a check at coin activation that if trezor is enabled then chain_id must be set in coins.
And this err apparently is not BAD_REQUEST but rather INTERNAL_SERVER_ERROR
(I can do this fix myself)

shamardy replied:
I thought it's better to make the `chain_id` non optional instead but found that there are 2 coins in coin config [BIDR-BEP20](https://github.com/KomodoPlatform/coins/blob/2e6e232994dfc11880cfe4de094144d09e709fec/coins#L1666) and [VRA-ERC20](https://github.com/KomodoPlatform/coins/blob/2e6e232994dfc11880cfe4de094144d09e709fec/coins#L12220) with no chain id, I think these 2 coins configs should be fixed
c.c. @smk762 @cipig

shamardy replied:
I will proceed with making `chain_id` non optional @dimxy

shamardy replied:
Fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/def7e33cc9bc56866452c7772b16601afb608c1c

dimxy commented on mm2src/coins/eth/eth_hd_wallet.rs:
```diff
@@ -0,0 +1,206 @@
+    where
+        XPubExtractor: HDXPubExtractor + Send,
+    {
+        extract_extended_pubkey(self, xpub_extractor, derivation_path).await
```
maybe good to rename this func to extract_extended_pubkey_impl() or extract_extended_pubkey_common_impl()

shamardy replied:
Fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/fb1ccd724af5ae2c6ec5fec672883c131d0441a5

dimxy commented on mm2src/coins/hd_wallet/pubkey.rs:
```diff
@@ -87,20 +26,23 @@ pub trait ExtractExtendedPubkey {
+    async fn extract_xpub(
         &self,
-        trezor_utxo_coin: String,
+        trezor_coin: String,
```
I guess this trait was not designed for trezor only so this `trezor_coin` probably will be wrapped in a enum when we add another device

shamardy replied:
Exactly. It should be refactored when adding support for other devices.

laruh commented on mm2src/coins/eth/eth_hd_wallet.rs:
```diff
@@ -0,0 +1,206 @@
+}
+
+impl EthCoin {
+    // Todo: make a common implementation for all coins
```
If we want to have a common impl, I suggest to create a trait type then and implement it.

shamardy replied:
We already have `CoinWithDerivationMethod` trait that can be used for this https://github.com/KomodoPlatform/komodo-defi-framework/blob/fb1ccd724af5ae2c6ec5fec672883c131d0441a5/mm2src/coins/lp_coins.rs#L3904
I fixed this here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/26cca4beb06c30e8a4534866d6763cd493dc66a3 but renamed it to all_addresses since there is another `my_addresses` method used for utxo transaction history. The two should be merged in the future by making utxo tx history implementation generic across all coins.

laruh commented on mm2src/coins/hd_wallet/coin_ops.rs:
```diff
@@ -0,0 +1,234 @@
+            .collect()
+    }
+
+    // Todo: combine both implementations once worker threads are supported in WASM
```
Could you remind me please, as far as I remember we added worker support in wasm, right?
Do we need something else for this todo?

shamardy replied:
Workers needs to be supported from the GUI side as well by adding a worker file for this and implementing any needed code if there is any.

shamardy commented on mm2src/coins/hd_wallet/pubkey.rs:
```diff
@@ -1,79 +1,18 @@
-}
-
 #[async_trait]
 pub trait ExtractExtendedPubkey {
```
Done here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/faaae10af3d8d25c9c8a7b1b62ed247b2d36dece

shamardy commented on mm2src/coins/eth/eth_balance_events.rs:
```diff
@@ -10,65 +11,104 @@ use mm2_err_handle::prelude::MmError;
 use super::EthCoin;
 use crate::{eth::{u256_to_big_decimal, Erc20TokenInfo},
             BalanceError, MmCoin};
+struct BalanceData {
```
Fixed here https://github.com/KomodoPlatform/komodo-defi-framework/commit/26cca4beb06c30e8a4534866d6763cd493dc66a3

dimxy commented on mm2src/coins_activation/src/eth_with_token_activation.rs:
```diff
@@ -35,10 +46,7 @@ impl From<EthActivationV2Error> for EnablePlatformCoinWithTokensError {
-            EthActivationV2Error::ExpectedRpcChainId => {
-                EnablePlatformCoinWithTokensError::InvalidPayload(err.to_string())
-            },
+            EthActivationV2Error::ChainIdNotSet => EnablePlatformCoinWithTokensError::InvalidPayload(err.to_string()),
```
I think 'InvalidPayload' error assumes that something wrong is with the activation request (as 'payload' is usually meant as the content of a message). In case of a missing chain id I'd say this is rather an internal error or invalid config error.

shamardy replied:
Fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/22013005986f908bda35bb21b07803d7f798e90a

I also added a checklist item
>  Document changes before merging PRs. e.g. chain_id is needed for all evm coins/tokens. chain_id should be part of platform coin config only actually and retrieved from platform coin.

To remember documenting that chain_id is now mandatory ref. https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979#discussion_r1546841562

Btw, I think `chain_id`, `required_confirmations`, `avg_blocktime`, `derivation_path` should be inherited from platform coin and no need to add them to every token config as it's currently done. There will be backward compatibility issues for v1 activation methods if this is done though, maybe I should open an issue.

shamardy replied:
> Btw, I think chain_id, required_confirmations, avg_blocktime, derivation_path should be inherited from platform coin and no need to add them to every token config as it's currently done. There will be backward compatibility issues for v1 activation methods if this is done though, maybe I should open an issue.

P.S. This can be part of custom token/asset import functionality which we will work on soon.

shamardy replied:
found an issue for this already for QRC20 https://github.com/KomodoPlatform/komodo-defi-framework/issues/756

shamardy commented:
@dimxy @laruh are you working on any more review iterations for this PR? If not can you please check if all comments are fixed and approve the PR so I can merge it.

shamardy commented:
@onur-ozkan just need you to check this change https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979#discussion_r1554473845 and if it's alright by you.

laruh reviewed (approved):
LGTM! 🔥

dimxy reviewed (approved):
LGTM
